src/storage/CustomArrayList.java

```diff
@@ -0,0 +1,163 @@
+        System.arraycopy(newArray, 0, array, index + 1, newArray.length);
+
+        elementsAmount++;
+    }
+
+    //new array will be added to the end of array
+    public void add(T[] elements) {
+
+        notNullCheck(array);
+        notNullCheck(elements);
+
+        while ((elementsAmount + elements.length) > array.length) {
+            resize();
+        }
+
+        System.arraycopy(elements, 0, array, elementsAmount, elements.length);
+        elementsAmount = elementsAmount + elements.length;
+    }
+
+    //new array will be inserted starting at the index cell
+    public void add(T[] elements, int startIndex) {
+
+        notNullCheck(array);
+        notNullCheck(elements);
+
+        if (startIndex < 0 || startIndex > elementsAmount) {
+            throw new IndexOutOfBoundsException("Index must be between 0 and " + elementsAmount + ", your index is " + startIndex);
+        }
+
+        while ((elementsAmount + elements.length) > array.length) {
+            resize();
+        }
+
+        Object[] newArray = new Object[elementsAmount - startIndex];
+
+        System.arraycopy(array, startIndex, newArray, 0, elementsAmount - startIndex);
+        System.arraycopy(elements, 0, array, startIndex, elements.length);
+        System.arraycopy(newArray, 0, array, startIndex+elements.length, newArray.length);
+
+        elementsAmount = elementsAmount + elements.length;
+    }
+
+    //the value of element will be put at the index cell, if index cell has value, it will be replaced by new value
+    public void put(T element, int index) {
+        notNullCheck(array);
+        notNullCheck(element);
+
+        if (index >= 0 && index < elementsAmount) {
+            array[index] = element;
+        } else
+            throw new IndexOutOfBoundsException("Index must be between 0 and " + (elementsAmount - 1) + ", your index is " + index);
+    }
+
+    //the array will be put, starting at the index cell, if index cell has value, it will be replaced by new value
+    public void put(T[] elements, int startIndex) {
+        notNullCheck(array);
+        notNullCheck(elements);
+
+        while ((startIndex + elements.length) > array.length) {
+            resize();
+        }
+
+        if (startIndex >= 0 && startIndex < elementsAmount) {
+            System.arraycopy(elements, 0, array, startIndex, elements.length);
+            elementsAmount = startIndex + elements.length;
+        } else
+            throw new IndexOutOfBoundsException("Index must be between 0 and " + (size() - 1) + ", your index is " + startIndex);
+
+    }
+
+    /*  TO GET BY INDEX */
+
+    public T getByIndex(int index) {
+        if (index >= 0 && index < elementsAmount) {
+            return (T) array[index];
+        } else
+            throw new IndexOutOfBoundsException("Index must be between 0 and " + (elementsAmount - 1) + ", your index is " + index);
+    }
+
+    /*  TO DELETE BY INDEX */
+
+    public void deleteByIndex(int index) {
```

Comment:
if you check `if (index >= 0 && index < elementsAmount` in `getByIndex`, then it is good approach to check it here as well

Reply:
fixed